File: design-patterns/golang/prototype/enemy_registry.go
package prototype

// EnemyRegistry manages enemy prototypes
type EnemyRegistry struct {
	prototypes map[string]*Enemy
}

// NewEnemyRegistry creates a new EnemyRegistry
func NewEnemyRegistry() *EnemyRegistry {
	return &EnemyRegistry{
		prototypes: make(map[string]*Enemy),
	}
}

// Register adds a prototype to the registry
func (r *EnemyRegistry) Register(key string, prototype *Enemy) {
	r.prototypes[key] = prototype
}

// Get retrieves and clones a prototype from the registry
func (r *EnemyRegistry) Get(key string) *Enemy {
	if prototype, exists := r.prototypes[key]; exists {
		return prototype.Clone().(*Enemy)
	}
	panic("No prototype registered for: " + key)
}

File: design-patterns/python/prototype/enemy_registry.py
from typing import Dict
from .enemy import Enemy


class EnemyRegistry:
    """Registry for managing enemy prototypes"""
    
    def __init__(self):
        self._prototypes: Dict[str, Enemy] = {}
    
    def register_enemy(self, key: str, enemy: Enemy) -> None:
        """Register an enemy prototype"""
        self._prototypes[key] = enemy
    
    def get_enemy(self, key: str) -> Enemy:
        """Get a cloned enemy from the registry"""
        if key not in self._prototypes:
            raise KeyError(f"Enemy type '{key}' not found in registry")
        return self._prototypes[key].clone()
    
    def get_available_types(self) -> list:
        """Get list of available enemy types"""
        return list(self._prototypes.keys())